packages/contracts/test/RTokenWithAave.test.js

```diff
@@ -0,0 +1,506 @@
+        const accounts = [admin, customer1, customer2, customer3, customer4];
+        let totalSupplyByAccounts = toWad(0);
+        let totalSavingsAmountByAccounts = toWad(0);
+        let totalReceivedLoansByAccounts = toWad(0);
+        let totalDebtFreeInterestByAccounts = toWad(0);
+
+        for (let i = 0; i < accounts.length; ++i) {
+            const account = accounts[i];
+            const stats = await rToken.getAccountStats.call(account);
+            totalSupplyByAccounts = totalSupplyByAccounts
+                .add(web3.utils.toBN(await rToken.balanceOf.call(account)));
+            totalSavingsAmountByAccounts = totalSavingsAmountByAccounts
+                .add(web3.utils.toBN(await rToken.receivedSavingsOf.call(account)));
+            totalReceivedLoansByAccounts = totalReceivedLoansByAccounts
+                .add(web3.utils.toBN(await rToken.receivedLoanOf.call(account)));
+            totalDebtFreeInterestByAccounts = totalDebtFreeInterestByAccounts
+                .add(web3.utils.toBN(stats.cumulativeInterest))
+                .sub(web3.utils.toBN(stats.rInterest));
+        }
+
+        const globalStats = await rToken.getGlobalStats.call();
+        assert.deepEqual({
+            totalSupply: totalSupplyByAccounts.toString(),
+            totalSavingsAmount: wad4human(totalSavingsAmountByAccounts, 12)
+        }, {
+            totalSupply: globalStats.totalSupply.toString(),
+            totalSavingsAmount: wad4human(globalStats.totalSavingsAmount, 12)
+        }, "invariants: accountStats vs globalStats");
+
+        const nHats = parseInt((await rToken.getMaximumHatID.call()).toString()) + 1;
+        let totalReceivedLoansByHats = toWad(0);
+        let totalSavingsByHats = toWad(0);
+        for (let i = 0; i <= nHats; ++i) {
+            let hatID = i;
+            if (i === nHats) hatID = SELF_HAT_ID;
+            const stats = await rToken.getHatStats.call(hatID);
+            totalReceivedLoansByHats = totalReceivedLoansByHats
+                .add(web3.utils.toBN(stats.totalLoans));
+            totalSavingsByHats = totalSavingsByHats
+                .add(web3.utils.toBN(stats.totalSavings));
+        }
+        assert.deepEqual({
+            totalReceivedLoans: totalReceivedLoansByAccounts.toString(),
+            totalSavings: wad4human(totalSavingsAmountByAccounts.add(totalDebtFreeInterestByAccounts), 6),
+        }, {
+            totalReceivedLoans: totalReceivedLoansByHats.toString(),
+            totalSavings: wad4human(totalSavingsByHats, 6),
+        }, "invariants: accountStats vs hatStats");
+    }
+
+    it("#0 initial test condition", async () => {
+        assert.equal(wad4human(await rToken.totalSupply.call()), "0.00000");
+        assert.equal(wad4human(await aToken.balanceOf.call(customer1)), "0.00000");
+        assert.equal(wad4human(await token.balanceOf.call(customer1)), "1000.00000");
+    });
+
+    it("#1 normal operations with zero hatter", async () => {
```

Comment:
Good stuff.

Rtoken team will continue from here, and refactor our test suite to test this strategy a bit more